### PR TITLE
Add initial performance shader support (matmul)

### PR DESCRIPTION
**Pull request by max-Hawkins**
This adds support for Metal Performance Shaders but only implements matrix-matrix multiplication for a few type combinations. However, it provides for easier MPS additions to be made in the future, so I decided to cut off a smaller PR. Hopefully it's now easy enough for others to add what they want/need.

There are still two test cases with mul! failing for Float16 and Float32 matrices with the identity operator applied to them. Unsure why these are failing. I also wasn't fully confident on how to dispatch from mul! to the MPS matmul kernel as there are some intricacies to it (needing alpha and beta to always be Float64 and the memory layout differences between MPSMatrix and Julia matrices). This latter point could be used to optimize away some of the transpose operations for certain input matrix sizes. @maleadt Any input?

### CODE REVIEW

File: src/device/performance_shaders/mps_matrix.jl [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+                                alpha, # alpha
+                                beta,  # beta
+                                )
+
+    cmdbuf = MtlCommandBuffer(global_queue(current_device()))
+
+    # Encode and commit matmul kernel
+    Metal.MTL.mtMPSMatMulEncodeToCommandBuffer(mat_mul_kernel, cmdbuf, mps_a, mps_b, mps_c)
+    commit!(cmdbuf)
+    synchronize()
```
Comment (maleadt) [inline src/device/performance_shaders/mps_matrix.jl:59]: We don't want to synchronize here. The transpose below should be encoded on the same queue.

File: src/linalg.jl [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        false, B
+    end
+
+    typA = eltype(A)
+    typB = eltype(B)
+    typC = eltype(C)
+
+    # If possible, dispatch to performance shaders
+    if typA == typB && (typA, typC) in Metal.MPS_VALID_MATMUL_TYPES
+        Metal.matMulMPS(C, dA, dB, Float64(alpha), Float64(beta), tA, tB)
```
Comment (maleadt) [inline src/linalg.jl:44]: No need to convert alpha/beta eagerly, `ccall` will do that for you.

File: deps/cmt/src/performance_shaders/mps_matrix.m [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+CF_RETURNS_RETAINED
+MT_EXPORT
+MT_API_AVAILABLE(mt_macos(10.13), mt_ios(10.0))
+MtMPSMatrixMultiplication*
+mtNewMPSMatrixMultiplication(MtDevice *device, bool transposeLeft, bool transposeRight,
+                             NsUInteger resultRows, NsUInteger resultColumns,
+                             NsUInteger interiorColumns, double alpha, double beta){
+
+return (MtMPSMatrix *)[[MPSMatrixMultiplication alloc] initWithDevice:(id<MTLDevice>)device
```
Comment (maleadt) [inline deps/cmt/src/performance_shaders/mps_matrix.m:46]: It's weird that the docs mention:

> alpha and beta are scalar values of the same data type as the values of C

File: src/device/performance_shaders/mps_matrix.jl [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+end
+
+"""
+
+matMulMPS(a::MtlMatrix, b::MtlMatrix, c::MtlMatrix, alpha=Float64(1), beta=Float64(1),
+transpose_left=false, transpose_right=false)
+
+Perform  c = alpha * op(a) * beta * op(b) + beta * C using Metal Performance Shaders
+"""
+function matMulMPS(c::MtlMatrix, a::MtlMatrix, b::MtlMatrix, alpha::Float64=Float64(1), beta::Float64=Float64(1),
```
Comment (maleadt) [inline src/device/performance_shaders/mps_matrix.jl:31]: I'd relax `alpha` and `beta`'s type signatures here, but use `::Number=true` instead; `ccall` will convert to Float64.

**Comment by maleadt** [comment]
> This latter point could be used to optimize away some of the transpose operations for certain input matrix sizes.

Can't we always use (AB)ᵀ = BᵀAᵀ?

**Comment by maleadt** [comment]
> needing alpha and beta to always be Float64

The docs state that the type of alpha/beta should match C, so maybe some truncation is happening?

**Comment by maleadt** [comment]
Should probably add some tests for the mixed-precision cases, as GPUArrays cannot assume that the back-end has one (and the generic matmatmul doesn't work in extended precision, although maybe it should...)

**Comment by maleadt** [comment]
Quick performance test:

```julia
julia> n = p = m = 2048
2048

julia> a = MtlArray(rand(Float32, n, p));

julia> b = MtlArray(rand(Float32, p, m));

julia> c = MtlArray(zeros(Float32, n, m));

julia> bench = @benchmark Metal.@sync mul!(c, a, b)
BenchmarkTools.Trial: 518 samples with 1 evaluation.
 Range (min … max):  9.366 ms …  13.354 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     9.629 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   9.646 ms ± 192.169 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

               ▃▂▅▅▆▆▆▇█▇▇▆▅▄▄▁▁ ▁
  ▄▁▄▄▄▄▆▆▆▄▄▁▇█████████████████▄█▄▁▆▁▄▁▆▁▇▁▄▄▁▁▄▄▇▁▄▆▄▁▁▁▁▁▄ █
  9.37 ms      Histogram: log(frequency) by time      10.1 ms <

 Memory estimate: 352 bytes, allocs estimate: 12.

julia> flops = n*m*(2p-1)
17175674880

julia> flops / (minimum(bench.times)/1e9)
1.8339303699749077e12
```

So 1.8 TFlops, using an 8-core (CPU and GPU) M1 which theoretically has a 2.6TFlops throughput. Pretty nice already!

File: test/device/performance_shaders.jl [hunk 5]
```diff
@@ -0,0 +1,39 @@
+
+@testset "Performance Shaders - MatMul" begin
```
Comment (maleadt) [inline test/device/performance_shaders.jl:2]: the `device` subdirs are typically about Julia device code, so I'd move the `performance_shaders` out just `test/mps.jl`.

**Comment by max-Hawkins** [comment]
Just pushed my mixed-precision tests that I had been using. The previously failing tests were due to overflowing the max/min values.

**Comment by max-Hawkins** [comment]
With this update, we've achieved > 70% of peak floating point performance. On my base (7 core GPU) M1 MacBook Air, the theoretical performance is 2.6 TFLOPS * 7/8 = 2.275 TFLOPS. The below results show achieving ~1.63 TFLOPS.

```
julia> bench_mps(N, Float32, Float32)
[ Info: Benchmarking MPS (N=8192  Float32 => Float32)...
[ Info: Running bench...
[ Info: TOPS: 1.628
BenchmarkTools.Trial: 40 samples with 1 evaluation.
 Range (min … max):  675.149 ms … 946.328 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     757.493 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   766.657 ms ±  67.500 ms  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▁   ▁ ▄ █▄            ▁▁        ▄    ▄                         
  █▆▁▁█▆█▁██▆▁▁▆▆▁▁▆▁▆▆▁██▁▁▁▆▁▆▁▁█▁▆▁▁█▁▆▁▁▁▆▆▆▁▁▁▁▁▁▁▁▁▁▁▁▁▁▆ ▁
  675 ms           Histogram: frequency by time          946 ms <

 Memory estimate: 352 bytes, allocs estimate: 12.
```

With Float16 -> Float32, I recorded 1.92 TFLOPS.

```
julia> bench_mps(N, Float16, Float32)
[ Info: Benchmarking MPS (N=8192  Float16 => Float32)...
[ Info: Running bench...
[ Info: TOPS: 1.917
BenchmarkTools.Trial: 50 samples with 1 evaluation.
 Range (min … max):  573.563 ms … 677.480 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     607.186 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   611.689 ms ±  23.636 ms  ┊ GC (mean ± σ):  0.00% ± 0.00%

     ▃   ▃ ▃▃  ▃▃▃ ▃▃   ▃         ▃▃  █▃▃                        
  ▇▇▇█▁▁▇█▇██▇▇███▁██▁▇▇█▁▁▁▁▇▁▇▇▇██▇▇███▇▇▁▁▁▁▇▁▁▁▁▁▁▁▁▁▁▁▁▁▁▇ ▁
  574 ms           Histogram: frequency by time          677 ms <

 Memory estimate: 352 bytes, allocs estimate: 12.
```

**Pull request merged**